ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/QueryResultImpl.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.client.impl;
+
+import io.confluent.ksql.api.client.QueryResult;
+import io.confluent.ksql.api.client.Row;
+import io.confluent.ksql.reactive.BufferedPublisher;
+import io.vertx.core.Context;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.reactivestreams.Subscriber;
+
+class QueryResultImpl extends BufferedPublisher<Row> implements QueryResult {
+
+  private final String queryId;
+  private final List<String> columnNames;
+  private final List<String> columnTypes;
+  private final PollableSubscriber pollableSubscriber;
+  private boolean polling;
+  private boolean subscribing;
+
+  QueryResultImpl(final Context context, final String queryId, final List<String> columnNames,
+      final List<String> columnTypes) {
+    super(context);
+    this.queryId = queryId;
+    this.columnNames = columnNames;
+    this.columnTypes = columnTypes;
+    this.pollableSubscriber = new PollableSubscriber(ctx);
+  }
+
+  @Override
+  public List<String> columnNames() {
+    return columnNames;
+  }
+
+  @Override
+  public List<String> columnTypes() {
+    return columnTypes;
+  }
+
+  @Override
+  public String queryID() {
+    return queryId;
+  }
+
+  @Override
+  public synchronized void subscribe(final Subscriber<? super Row> subscriber) {
+    if (polling) {
+      throw new IllegalStateException("Cannot set subscriber if polling");
+    }
+    subscribing = true;
+    super.subscribe(subscriber);
+  }
+
+  @Override
+  public Row poll() {
+    return poll(0, TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public synchronized Row poll(final long timeout, final TimeUnit timeUnit) {
+    if (subscribing) {
+      throw new IllegalStateException("Cannot poll if subscriber has been set");
+    }
+    if (!polling) {
+      subscribe(pollableSubscriber);
+      subscribing = false;
+      polling = true;
+    }
+    try {
+      return pollableSubscriber.poll(timeout, timeUnit);
+    } catch (final Throwable t) {
+      sendError(new Exception("Failed to poll", t));
+      return null;
+    }
+  }
+
+  @Override
+  public boolean isComplete() {
+    return false;
```

Comment:
Always returns false?

Reply:
Whoops! Forgot to implement this -- thanks for the catch.

Now that I think about it, though, I'm unclear on what the desired behavior is. Does "complete" mean that the underlying connection is closed, or that all rows have been read? I see that for BufferedPublisher "complete" means the former, as it's possible for the BufferedPublisher to be "complete" even if there are buffered rows that have yet to be read, but this seems a little counterintuitive.

Reply:
Complete for a subscriber means the onComplete method has been called.
onComplete should be called for any graceful close - termination by server, end of limit, or close by the close-query api.
The complete method on a publisher means "mark this publisher as complete, and send onComplete to it's subscribers once you've delivered any pending rows"

Reply:
Got'cha. I've renamed a couple internal variables in `BufferedPublisher` to better reflect this. Thanks for the clarification!